src/librustc_codegen_ironox/value.rs

```diff
@@ -28,22 +33,54 @@ pub enum Value {
     ConstInt(usize),
     /// The index of an `IronOxStruct` in the `structs` vec from `ModuleIronOx`.
     ConstStruct(usize),
-    /// The parameter of an `IronOxFunction`. This is just a wrapper around a `Type`.
-    Param(Type),
+    /// The parameter of an `IronOxFunction`. This is just a wrapper around a
+    /// `Type`. A parameter is an (index, type) pair, where 'index' is the
+    /// index of the parameter in the list of parameters of the function.
+    Param(usize, Type),
+    /// An instruction: (functiton index, basic block index, instruction index).
+    Instruction(usize, usize, usize),
     /// A placeholder for unimplemented Values. This variant will be removed.
     None,
 }
 
-#[derive(PartialEq, Copy, Clone, Debug)]
+/// An IronOx instruction.
+#[derive(PartialEq, Debug, Eq, Hash)]
 pub enum Instruction {
-    // FIXME: implement
+    /// Call(fn_idx, args). Emit a call to the function found at index `fn_idx`
+    /// in the `functions` vector of the module.
+    Call(usize, Vec<Value>),
+    /// Allocate space on the stack for a variable of a particular type and
+    /// alignment.
+    Alloca(String, Type, Align),
+    /// Cast a value to a type.
+    Cast(Value, Type),
+    /// Add two values and return the result.
+    Add(Value, Value),
+    /// Currently serves as a no-op. This variant will be removed.
     None,
 }
 
-impl Eq for Value {}
-
-impl Hash for Value {
-    fn hash<H: Hasher>(&self, hasher: &mut H) {
-        (self as *const Self).hash(hasher);
+impl Instruction {
+    pub fn val_ty(&self, cx: &CodegenCx, module: &ModuleIronOx) -> Type {
+        match *self {
+            Instruction::Alloca(_, ty, _) => {
+                ty
+            },
+            Instruction::Cast(_, ty) => {
+                ty
+            },
+            Instruction::Add(v1, v2) => {
+                let ty1 = cx.val_ty(v1);
+                let ty2 = cx.val_ty(v2);
+                assert_eq!(ty1, ty2);
+                ty1
+            },
+            Instruction::Call(fn_idx, _) => {
+                cx.val_ty(cx.module.borrow().functions[fn_idx].ret)
+            },
+            _ => {
```

Comment:
Is this branch necessary? In other words, if you're catering for all the current `Instruction::` variants above, it would be better to remove this clause, then you'll get a static error when something new is added to `Instruction.

Reply:
I guess not, it's just that I still have an `Instruction::None` which I don't want to handle (it's sort of a placeholder, but it's removed in my next PR). Either way, I removed the branch.

Reply:
OK. It's probably best to say `Instruction::None => unimplemented()` and remove it in your next PR?

Reply:
Fixed.